android/pytorch_android/src/main/cpp/pytorch_jni.cpp

```diff
@@ -0,0 +1,617 @@
+      inputDataCapacity);
+  return inputTensor;
+}
+
+class JTensor : public facebook::jni::JavaClass<JTensor> {
+ public:
+  constexpr static const char* kJavaDescriptor = "Lorg/pytorch/Tensor;";
+
+  static facebook::jni::local_ref<JTensor> newJTensor(
+      facebook::jni::alias_ref<facebook::jni::JByteBuffer> jBuffer,
+      facebook::jni::alias_ref<jlongArray> jDims,
+      jint typeCode) {
+    static auto jMethodNewTensor =
+        JTensor::javaClassStatic()
+            ->getStaticMethod<facebook::jni::local_ref<JTensor>(
+                facebook::jni::alias_ref<facebook::jni::JByteBuffer>,
+                facebook::jni::alias_ref<jlongArray>,
+                jint)>("nativeNewTensor");
+    return jMethodNewTensor(
+        JTensor::javaClassStatic(), jBuffer, jDims, typeCode);
+  }
+
+  static facebook::jni::local_ref<JTensor> newJTensorFromAtTensor(
+      const at::Tensor& tensor) {
+    const auto scalarType = tensor.scalar_type();
+    int typeCode = 0;
+    if (at::kFloat == scalarType) {
+      typeCode = kTensorTypeCodeFloat32;
+    } else if (at::kInt == scalarType) {
+      typeCode = kTensorTypeCodeInt32;
+    } else if (at::kByte == scalarType) {
+      typeCode = kTensorTypeCodeByte;
+    } else {
+      facebook::jni::throwNewJavaException(
+          facebook::jni::gJavaLangIllegalArgumentException,
+          "at::Tensor scalar type is not supported on java side");
+    }
+
+    const auto& tensorDims = tensor.sizes();
+    std::vector<int64_t> tensorDimsVec;
+    for (const auto& dim : tensorDims) {
+      tensorDimsVec.push_back(dim);
+    }
+
+    facebook::jni::local_ref<jlongArray> jTensorDims =
+        facebook::jni::make_long_array(tensorDimsVec.size());
+
+    jTensorDims->setRegion(0, tensorDimsVec.size(), tensorDimsVec.data());
+
+    facebook::jni::local_ref<facebook::jni::JByteBuffer> jTensorBuffer =
+        facebook::jni::JByteBuffer::allocateDirect(tensor.nbytes());
+    jTensorBuffer->order(facebook::jni::JByteOrder::nativeOrder());
+    std::memcpy(
+        jTensorBuffer->getDirectBytes(),
+        tensor.storage().data(),
+        tensor.nbytes());
+    return JTensor::newJTensor(jTensorBuffer, jTensorDims, typeCode);
+  }
+
+  static at::Tensor newAtTensorFromJTensor(
+      facebook::jni::alias_ref<JTensor> jtensor) {
+    static const auto typeCodeMethod =
+        JTensor::javaClassStatic()->getMethod<jint()>("getTypeCode");
+    jint typeCode = typeCodeMethod(jtensor);
+
+    static const auto dimsField =
+        JTensor::javaClassStatic()->getField<jlongArray>("dims");
+    auto jdims = jtensor->getFieldValue(dimsField);
+
+    static auto dataBufferMethod =
+        JTensor::javaClassStatic()
+            ->getMethod<
+                facebook::jni::local_ref<facebook::jni::JBuffer::javaobject>()>(
+                "getRawDataBuffer");
+    facebook::jni::local_ref<facebook::jni::JBuffer> jbuffer =
+        dataBufferMethod(jtensor);
+    return newAtTensor(jbuffer, jdims, typeCode);
+  }
```

Comment:
Java n00b question: is there a specific reason we don't let `org.pytorch.Tensor` hold `at::Tensor` on JNI side, and instead rely on the conversion when we need the `at::Tensor` value?